### PR TITLE
Text recognition crnn

### PR DESCRIPTION
**Pull request by LaurentBerger**
new c++ sample https://github.com/opencv/opencv_zoo/tree/main/models/text_recognition_crnn

### CODE REVIEW

File: models/text_recognition_crnn/demo.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
+#include <iostream>
+#include <uchar.h>
```
Comment (fengyuentau) [inline models/text_recognition_crnn/demo.cpp:2]: use `<cuchar>` since it is c++?

**Comment by WanliZhong** [comment]
@LaurentBerger Hi, thanks for contribution, this PR need to update readme document.

File: models/text_recognition_crnn/README.md [hunk 2]
```diff
@@ -59,6 +61,31 @@ python demo.py --input /path/to/image --model text_recognition_CRNN_CN_2021nov.o
...
 ```
+### C++
+
+Install latest OpenCV and CMake >= 3.24.0 to get started with:
+
+```shell
+# detect on camera input
+./build/opencv_zoo_text_recognition_crnn 
+# detect on an image
+./build/opencv_zoo_text_recognition_crnn --input /path/to/image -v
```
Comment (WanliZhong) [inline models/text_recognition_crnn/README.md:72]: need to use `-i=/path/to/image` or `--input=/path/to/image`

File: models/text_recognition_crnn/README.md [hunk 3]
```diff
@@ -59,6 +61,31 @@ python demo.py --input /path/to/image --model text_recognition_CRNN_CN_2021nov.o
 # get help regarding various parameters
 python demo.py --help
 ```
+### C++
+
+Install latest OpenCV and CMake >= 3.24.0 to get started with:
+
+```shell
+# detect on camera input
+./build/opencv_zoo_text_recognition_crnn
```
Comment (WanliZhong) [inline models/text_recognition_crnn/README.md:70]: this command can't recognize from camera.

File: models/text_recognition_crnn/README.md [hunk 4]
```diff
@@ -59,6 +61,31 @@ python demo.py --input /path/to/image --model text_recognition_CRNN_CN_2021nov.o
...
+./build/opencv_zoo_text_recognition_crnn --help
+```
+
+Run the demo detecting Chinese:
+
+```shell
+# detect on camera input
+./build/opencv_zoo_text_recognition_crnn  --model text_recognition_CRNN_CN_2021nov.onnx --charset charset_3944_CN.txt
+# detect on an image
+./build/opencv_zoo_text_recognition_crnn  --input /path/to/image --model text_recognition_CRNN_CN_2021nov.onnx --charset charset_3944_CN.txt
```
Comment (WanliZhong) [inline models/text_recognition_crnn/README.md:84]: need to use `=`

File: models/text_recognition_crnn/demo.cpp [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        std::make_pair<cv::dnn::Backend, cv::dnn::Target>(dnn::DNN_BACKEND_CUDA, dnn::DNN_TARGET_CUDA),
+        std::make_pair<cv::dnn::Backend, cv::dnn::Target>(dnn::DNN_BACKEND_CUDA, dnn::DNN_TARGET_CUDA_FP16),
+        std::make_pair<cv::dnn::Backend, cv::dnn::Target>(dnn::DNN_BACKEND_TIMVX, dnn::DNN_TARGET_NPU),
+        std::make_pair<cv::dnn::Backend, cv::dnn::Target>(dnn::DNN_BACKEND_CANN, dnn::DNN_TARGET_NPU)};
+        
+vector<u16string> loadCharset(string);
+
+std::string keys =
+"{ help  h           |                                              | Print help message. }"
+"{ model m           | text_detection_DB_IC15_resnet18_2021sep.onnx | Usage: Set model type, defaults to text_detection_DB_IC15_resnet18_2021sep.onnx }"
```
Comment (WanliZhong) [inline models/text_recognition_crnn/demo.cpp:26]: Default model should be text recognize model, not detection model. Or, If we can add another parameter to set detection model and recognize model separately?

**Review by WanliZhong — APPROVED** [review]
Thanks for your contribution!👍👍👍, here are something I think need to be modified.

File: models/text_recognition_crnn/README.md (hunk 2 above)
Comment (WanliZhong) [reply models/text_recognition_crnn/README.md:72]: using `-v` flag can't show the image

Comment (LaurentBerger) [reply models/text_recognition_crnn/README.md:72]: ./build/opencv_zoo_text_recognition_crnn -v=0
show result as text but won't stop without ctrl+c. Must I add a counter to stop after 100 inferences?

Comment (WanliZhong) [reply models/text_recognition_crnn/README.md:72]: I think just be as consistent as possible with other demos

Comment (fengyuentau) [reply models/text_recognition_crnn/README.md:72]: > using -v flag can't show the image

Just modify the doc here to make sure the commands work. It does not worth making 100% same behavior as Python demos.

File: models/text_recognition_crnn/README.md [hunk 6]
```diff
@@ -59,6 +61,31 @@ python demo.py --input /path/to/image --model text_recognition_CRNN_CN_2021nov.o
...
+./build/opencv_zoo_text_recognition_crnn --help
+```
+
+Run the demo detecting Chinese:
+
+```shell
+# detect on camera input
+./build/opencv_zoo_text_recognition_crnn  --model=text_recognition_CRNN_CN_2021nov.onnx --charset=charset_3944_CN.txt
+# detect on an image
+./build/opencv_zoo_text_recognition_crnn  --input=/path/to/image --model=text_recognition_CRNN_CN_2021nov.onnx --charset=charset_3944_CN.txt
```
[the remark this answers is not in the document]
Comment (LaurentBerger) [reply models/text_recognition_crnn/README.md:84]: May be there is something I don't understand : in readme 
```
Selected model must match with the charset:

    Try text_recognition_CRNN_EN_2021sep.onnx with charset_36_EN.txt.
    Try text_recognition_CRNN_CH_2021sep.onnx with charset_94_CH.txt
    Try text_recognition_CRNN_CN_2021sep.onnx with charset_3944_CN.txt.

```

Then  i understand that model gives charset to use then in code I use those line 


```
        if (this->modelPath.find("_EN_") != string::npos)
            this->charset = loadCharset("CHARSET_EN_36");
        else if (this->modelPath.find("_CH_") != string::npos)
            this->charset = loadCharset("CHARSET_CH_94");
        else if (this->modelPath.find("_CN_") != string::npos)
            this->charset = loadCharset("CHARSET_CN_3944");
        else
        {
            CV_Error(-1, "Charset not supported! Exiting ...");

        }

```

[the remark this answers is not in the document]
Comment (fengyuentau) [reply models/text_recognition_crnn/README.md:84]: We integrated charset characters into the class wrapper `crnn.py` from separate files, but forgot to update the information in readme. So the current readme is outdated, you might update in this pull request as well.

**Review by fengyuentau — APPROVED** [review]
LGTM 👍

**Pull request merged**